juju/juju
# Move retrystrategy facade to model config service

This PR refactors the agent/retrystrategy facade to use a model config service instead of state.
- The `st`, `model` and `resources` fields of `RetryStrategyAPI` were removed and replaced with `modelConfigService` and `watcherRegistry` as appropriate.
- Calls to `model.ModelConfig` were replaced by calls to `modelConfigService.ModelConfig`.
- Calls to `model.WatchForModelConfigChanges` were replaced by calls to `modelConfigService.Watch`, with the logic changed appropriately to handle the new watcher type.
- A `service.go` file was added defining the `interface ModelConfigService`.
- We generated a mock for `ModelConfigService` using gomock. The test suite was changed so that we no longer use `ApiServerSuite`, instead we use the generated mock.
- A new `NewRetryStrategyAPI` method was create to help enable testing.
- The `export_test.go` file was deleted as it became unnecessary.
- The `pacakge_test.go` was renamed to `package_test.go`.
- Drive-by: remove a code block in the uniter that was killing the leadership tracker worker and causing container crashes. This was only relevant for podspec charms and should be removed now. See https://github.com/juju/juju/pull/17484

## Checklist




## QA steps

Deploy an application, change the value of the `automatically-retry-hooks` model config key, and watch debug-log to see the retrystrategy / uniter workers restart with the new config.

```bash
juju bootstrap minikube c
juju add-model m
# set logging config on the model
juju model-config logging-config="<root>=WARNING;juju.worker.retrystrategy=DEBUG;juju.worker.uniter=DEBUG"
# deploy an application, any application
juju deploy tiny-bash
```

Toggle the value of `automatically-retry-hooks`:
```console
$ juju model-config automatically-retry-hooks
True
$ juju model-config automatically-retry-hooks=False
```

Watch debug-log on the model to see the uniter restart. This should be instant.
```
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.retrystrategy bouncing retrystrategy worker to get new values
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter juju-exec listener stopping
unit-tiny-bash-0: 11:46:31 INFO juju.worker.uniter unit "tiny-bash/0" shutting down: catacomb 0xc0004ad408 is dying
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter juju-exec listener stopped
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter starting local juju-exec listener on {unix /var/lib/juju/agents/unit-tiny-bash-0/run.socket <nil>}
unit-tiny-bash-0: 11:46:31 INFO juju.worker.uniter unit "tiny-bash/0" started
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter juju-exec listener running
unit-tiny-bash-0: 11:46:31 INFO juju.worker.uniter hooks are retried true
...
```

## Links



**Jira card:** JUJU-6098
